### PR TITLE
types: fix a bug in casting str2str when `union`

### LINKED ISSUE #31678: Truncation of UTF8mb4 values when the UNION statement for binary and non-binary charsets
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

Do  --binary-as-hex=false for better output
```sql
CREATE TABLE t1 (c VARCHAR(11)) CHARACTER SET utf8mb4;
CREATE TABLE t2 (b CHAR(1) CHARACTER SET binary, i INT);
INSERT INTO t1 (c) VALUES ('н1234567890');
INSERT INTO t2 (b, i) VALUES ('1', 1);
SELECT c FROM t1 UNION SELECT b FROM t2;
SELECT c FROM t1 UNION SELECT i FROM t2;
SELECT i FROM t2 UNION SELECT c FROM t1;
SELECT b FROM t2 UNION SELECT c FROM t1;

DROP TABLE t1, t2;

```
### 2. What did you expect to see? (Required)
```sql
mysql> CREATE TABLE t1 (c VARCHAR(11)) CHARACTER SET utf8mb4;
Query OK, 0 rows affected (0.01 sec)

mysql> CREATE TABLE t2 (b CHAR(1) CHARACTER SET binary, i INT);
Query OK, 0 rows affected (0.01 sec)

mysql> INSERT INTO t1 (c) VALUES ('н1234567890');
Query OK, 1 row affected (0.00 sec)

mysql> INSERT INTO t2 (b, i) VALUES ('1', 1);
Query OK, 1 row affected (0.00 sec)

mysql> SELECT c FROM t1 UNION SELECT b FROM t2;
+--------------+
| c            |
+--------------+
| н1234567890  |
| 1            |
+--------------+
2 rows in set (0.00 sec)

mysql> SELECT c FROM t1 UNION SELECT i FROM t2;
+--------------+
| c            |
+--------------+
| н1234567890  |
| 1            |
+--------------+
2 rows in set (0.00 sec)

mysql> SELECT i FROM t2 UNION SELECT c FROM t1;
+--------------+
| i            |
+--------------+
| 1            |
| н1234567890  |
+--------------+
2 rows in set (0.00 sec)

mysql> SELECT b FROM t2 UNION SELECT c FROM t1;
+--------------+
| b            |
+--------------+
| 1            |
| н1234567890  |
+--------------+
2 rows in set (0.01 sec)
```
### 3. What did you see instead (Required)
```sql

mysql> CREATE TABLE t1 (c VARCHAR(11)) CHARACTER SET utf8mb4;
Query OK, 0 rows affected (0.01 sec)

mysql> CREATE TABLE t2 (b CHAR(1) CHARACTER SET binary, i INT);
Query OK, 0 rows affected (0.01 sec)

mysql> INSERT INTO t1 (c) VALUES ('н1234567890');
Query OK, 1 row affected (0.00 sec)

mysql> INSERT INTO t2 (b, i) VALUES ('1', 1);
Query OK, 1 row affected (0.00 sec)

mysql> SELECT c FROM t1 UNION SELECT b FROM t2;
+-------------+
| c           |
+-------------+
| н123456789  |
| 1           |
+-------------+
2 rows in set, 1 warning (0.02 sec)

mysql> SELECT c FROM t1 UNION SELECT i FROM t2;
+--------------+
| c            |
+--------------+
| н1234567890  |
| 1            |
+--------------+
2 rows in set (0.00 sec)

mysql> SELECT i FROM t2 UNION SELECT c FROM t1;
+--------------+
| i            |
+--------------+
| н1234567890  |
| 1            |
+--------------+
2 rows in set (0.01 sec)

mysql> SELECT b FROM t2 UNION SELECT c FROM t1;
+-------------+
| b           |
+-------------+
| н123456789  |
| 1           |
+-------------+
2 rows in set, 1 warning (0.00 sec)

mysql> show warnings;
+---------+------+------------------------------------------+
| Level   | Code | Message                                  |
+---------+------+------------------------------------------+
| Warning | 1406 | Data Too Long, field len 11, data len 12 |
+---------+------+------------------------------------------+
1 row in set (0.00 sec)

```
### 4. What is your TiDB version? (Required)

```sql
| Release Version: v5.5.0-alpha-60-g8c70432b0-dirty
Edition: Community
Git Commit Hash: 8c70432b000d648e9ea26cc57a7fe79d2e67d341
Git Branch: master
UTC Build Time: 2022-01-07 17:47:06
GoVersion: go1.16.9
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false | 
```

### PR DESCRIPTION
**Pull request by CbcWestwolf**
### What problem does this PR solve?

Issue Number: close #31678

Problem Summary:

When building the plan of union, the `flen` of the projection target `FieldType` is mis-computed.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Fix a wrong casting in building union plan
```

### CODE REVIEW

**Comment by bb7133** [comment]
As you might noticed that the mysql-test case has been failed because of this PR(which is exactly the case from #31678), so please update the mysql-test case accordingly.

**Comment by CbcWestwolf** [comment]
> please update the mysql-test case accordingly

Sure! :-)

File: executor/executor_issue_test.go [hunk 1]
```diff
@@ -392,6 +392,38 @@ func TestIssue30971(t *testing.T) {
...
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("use test")
+
+	tk.MustExec("drop table if exists t1, t2;")
+	tk.MustExec("CREATE TABLE t1 (c VARCHAR(11)) CHARACTER SET utf8mb4")
+	tk.MustExec("CREATE TABLE t2 (b CHAR(1) CHARACTER SET binary, i INT)")
+	tk.MustExec("INSERT INTO t1 (c) VALUES ('н1234567890')")
+	tk.MustExec("INSERT INTO t2 (b, i) VALUES ('1', 1)")
+
+	var queries = []string{
```
Comment (xiongjiwei) [inline executor/executor_issue_test.go:407]: can be structured
```go
struct {
  query string
  expected int
}
```

File: executor/executor_issue_test.go [hunk 2]
```diff
@@ -392,6 +392,38 @@ func TestIssue30971(t *testing.T) {
...
 }
 
+func TestIssue31678(t *testing.T) {
+	// https://github.com/pingcap/tidb/issues/31678
+	store := testkit.CreateMockStore(t)
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("use test")
+
+	tk.MustExec("drop table if exists t1, t2;")
+	tk.MustExec("CREATE TABLE t1 (c VARCHAR(11)) CHARACTER SET utf8mb4")
```
Comment (xiongjiwei) [inline executor/executor_issue_test.go:402]: please add some other charset

**Review by xiongjiwei — APPROVED** [review]

**Pull request merged**